### PR TITLE
token_classification dataset modified to work with any token classification task

### PR DESCRIPTION
**Pull request by ekmb**
Single data layer for all token classification tasks, for example, NER, punctuation

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/modules/losses.py [hunk 1]
```diff
@@ -120,12 +120,16 @@ def __init__(self, num_classes, **kwargs):
         self._criterion = nn.CrossEntropyLoss()
         self.num_classes = num_classes
 
-    def _loss_function(self, logits, labels, input_mask):
-        active_loss = input_mask.view(-1) > 0.5
+    def _loss_function(self, logits, labels, loss_mask):
+        active_loss = loss_mask.view(-1)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/modules/losses.py]: why did we remove "> 0.5"

Comment (ekmb) [reply collections/nemo_nlp/nemo_nlp/modules/losses.py]: because we're now using loss_mask which contains True/False not 1/0 as the input_mask

Comment (chiphuyen) [reply collections/nemo_nlp/nemo_nlp/modules/losses.py]: that makes sense. we'd update the docstring to explain the data type each input takes.

File: collections/nemo_nlp/nemo_nlp/modules/losses.py [hunk 2]
```diff
@@ -120,12 +120,16 @@ def __init__(self, num_classes, **kwargs):
...
         active_logits = logits.view(-1, self.num_classes)[active_loss]
         active_labels = labels.view(-1)[active_loss]
 
-        loss = self._criterion(active_logits, active_labels)
+        # To support empty active_labels
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/modules/losses.py]: can you give examples of scenarios in which we might have empty active_labels?

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py [hunk 3]
```diff
@@ -1,58 +1,76 @@
...
 
     for kv, v in tensors.items():
-        if 'logits' in kv:
+        if kv.startswith('logits'):
             for v_tensor in v:
                 for logit_tensor in v_tensor:
-                    logits_lists.append(logit_tensor.detach().cpu().tolist())
+                    all_logits.append(tensor2list(logit_tensor))
+
+        if kv.startswith('labels'):
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: shouldn't this be elif?

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py [hunk 4]
```diff
@@ -1,58 +1,76 @@
...
+    global_vars["all_subtokens_mask"].extend(all_subtokens_mask)
 
-    global_vars["correct_labels"] += correct_labels
-    global_vars["incorrect_labels"] += incorrect_labels
-    global_vars["correct_preds"] += correct_preds
-    global_vars["total_preds"] += total_preds
-    global_vars["total_correct"] += total_correct
 
+def list2str(l):
+    return ' '.join([str(j) for j in l])
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: tensor2list and list2str should be in the common utils file since they can be reused for many cases

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py [hunk 5]
```diff
@@ -1,58 +1,76 @@
...
-    p = correct_preds / total_preds
-    r = correct_preds / total_correct
-    f1 = 2 * p * r / (p + r)
+    accuracy = sum(labels == preds) / labels.shape[0]
+    logger.info(f'Accuracy: {accuracy}')
 
-    print(f"Accuracy = {accuracy}")
-    print(f"F1= {f1}")
+    i = 0
+    if preds.shape[0] > 21:
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: where does 21 come from?

File: docs/sources/source/nlp/ner.rst [hunk 6]
```diff
@@ -12,115 +12,120 @@ This tutorial explains how to implement named entity recognition (NER) in NeMo.
...
 .. code-block::
 
-    Jennifer	B-PER
-    is		O
-    from	O
-    New		B-LOC
-    York	I-LOC
-    City	I-LOC
-    .		O
+    Jennifer is from New York City.
```
Comment (chiphuyen) [inline docs/sources/source/nlp/ner.rst]: shouldn't text in text.txt be tokenized? e.g. "City." should be "City ."?

File: docs/sources/source/nlp/ner.rst [hunk 7]
```diff
@@ -12,115 +12,120 @@ This tutorial explains how to implement named entity recognition (NER) in NeMo.
...
 
 Training
 --------
 
 .. tip::
 
     We recommend you try this out in a Jupyter notebook. It'll make debugging much easier!
+    See examples/nlp/NERWithBERT.ipynb
 
 First, we need to create our neural factory with the supported backend. How you should define it depends on whether you'd like to multi-GPU or mixed-precision training. This tutorial assumes that you're training on one GPU, without mixed precision. If you want to use mixed precision, set ``amp_opt_level`` to ``O1`` or ``O2``.
```
Comment (chiphuyen) [inline docs/sources/source/nlp/ner.rst]: explain that if we don't want to use mixed-precision, use ``O0``

File: docs/sources/source/nlp/ner.rst [hunk 8]
```diff
@@ -12,115 +12,120 @@ This tutorial explains how to implement named entity recognition (NER) in NeMo.
...
 
         nf = nemo.core.NeuralModuleFactory(backend=nemo.core.Backend.PyTorch,
-                                           local_rank=args.local_rank,
-                                           optimization_level=args.amp_opt_level,
-                                           log_dir=work_dir,
-                                           create_tb_writer=True,
-                                           files_to_copy=[__file__])
+                                           local_rank=None,
+                                           optimization_level=nemo.core.Optimization.mxprO0,
+                                           create_tb_writer=True)
```
Comment (chiphuyen) [inline docs/sources/source/nlp/ner.rst]: nemo.core.Optimization.mxprO0 might be already outdated. don't we use 'O0', 'O1', 'O2' now?

File: docs/sources/source/nlp/ner.rst [hunk 9]
```diff
@@ -12,115 +12,120 @@ This tutorial explains how to implement named entity recognition (NER) in NeMo.
...
 
     .. code-block:: python
 
-        tokenizer = NemoBertTokenizer(args.pretrained_bert_model)
-        pretrained_bert_model = nemo_nlp.huggingface.BERT(
-            pretrained_model_name=args.pretrained_bert_model)
+        tokenizer = NemoBertTokenizer(pretrained_model="bert-base-cased")
+        bert_model = nemo_nlp.huggingface.BERT(
+            pretrained_model_name="bert-base-cased",
+            factory=neural_factory)
```
Comment (chiphuyen) [inline docs/sources/source/nlp/ner.rst]: factory=neural_factory is no longer necessary

File: examples/nlp/token_classification.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                    mode='train',
+                    ignore_extra_tokens=args.ignore_extra_tokens,
+                    ignore_start_end=args.ignore_start_end,
+                    use_cache=args.use_cache):
+
+    nf.logger.info(f"Loading {mode} data...")
+    shuffle = args.shuffle_data if mode == 'train' else False
+
+    text_file = f'{args.data_dir}/text_{mode}.txt'
+    label_file = f'{args.data_dir}/labels_{mode}.txt'
```
Comment (chiphuyen) [inline examples/nlp/token_classification.py]: check to make sure these files exist. warn users if they don't.

File: examples/nlp/token_classification_infer.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+parser.add_argument("--add_brackets", action='store_false',
+                    help="Whether to take predicted label in brackets or \
+                    just append to word in the output")
+parser.add_argument("--work_dir", default='output/checkpoints', type=str)
+parser.add_argument("--labels_dict", default='label_ids.csv', type=str)
+parser.add_argument("--amp_opt_level", default="O0",
+                    type=str, choices=["O0", "O1", "O2"])
+
+args = parser.parse_args()
+print(args)
```
Comment (chiphuyen) [inline examples/nlp/token_classification_infer.py]: is it for debug? we can just copy the current file into the work_dir with files_to_copy=[__file__] in NeuralModuleFactory

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py (hunk 3 above)
Comment (ekmb) [reply collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: fixed

File: collections/nemo_nlp/nemo_nlp/modules/losses.py (hunk 2 above)
Comment (ekmb) [reply collections/nemo_nlp/nemo_nlp/modules/losses.py]: no, removed the empty label support

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py (hunk 4 above)
Comment (ekmb) [reply collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: done

File: collections/nemo_nlp/nemo_nlp/modules/losses.py (hunk 1 above)
Comment (ekmb) [reply collections/nemo_nlp/nemo_nlp/modules/losses.py]: updated the docstring

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py (hunk 5 above)
Comment (ekmb) [reply collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: added comments and introduced a variable for the magic number

File: examples/nlp/token_classification.py (hunk 10 above)
Comment (ekmb) [reply examples/nlp/token_classification.py]: done

File: examples/nlp/token_classification_infer.py (hunk 11 above)
Comment (ekmb) [reply examples/nlp/token_classification_infer.py]: I'm removing it from here, but leaving nf.logger.info(args) in the training script, as it could be useful for debugging. Saving the current file won't store args specified in the command line.

File: docs/sources/source/nlp/ner.rst (hunk 6 above)
Comment (ekmb) [reply docs/sources/source/nlp/ner.rst]: fixed

File: docs/sources/source/nlp/ner.rst (hunk 7 above)
Comment (ekmb) [reply docs/sources/source/nlp/ner.rst]: done

File: docs/sources/source/nlp/ner.rst (hunk 8 above)
Comment (ekmb) [reply docs/sources/source/nlp/ner.rst]: done

File: docs/sources/source/nlp/ner.rst (hunk 9 above)
Comment (ekmb) [reply docs/sources/source/nlp/ner.rst]: done

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py [hunk 12]
```diff
@@ -1,58 +1,99 @@
...
+        i = random.randint(0, preds.shape[0] - sample_size - 1)
+    logger.info("Sampled preds: [%s]" % list2str(preds[i:i+sample_size]))
+    logger.info("Sampled labels: [%s]" % list2str(labels[i:i+sample_size]))
+    logger.info(classification_report(labels, preds, target_names=label_ids))
 
-    p = correct_preds / total_preds
-    r = correct_preds / total_correct
-    f1 = 2 * p * r / (p + r)
+    # calculate and plot confusion_matrix
+    if graph_fold:
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/utils/callbacks/token_classification.py]: i think we should have an option in the argument to choose whether to show confusion_matrix. technically you can still show confusion matrix even if you don't have graph_fold. You only need graph_fold if you want to save the image.
